dev/generate_matrix.py

```diff
@@ -0,0 +1,209 @@
+    ['0.19.2', '0.20.4', '0.21.3', '0.22.2', '0.23.2']
+    """
+    versions = get_released_versionos(package_name)
+
+    # prevent passing non-existent versions
+    assert min_ver in versions
+    assert max_ver in versions
+
+    if excludes:
+        assert all(v in versions for v in excludes)
+        versions = filter(lambda v: v not in excludes, versions)
+
+    versions = filter(is_valid_version, versions)
+    versions = filter(lambda v: is_between(v, min_ver, max_ver), versions)
+    return list(_get_latest_micro_versions(versions))
+
+
+def get_changed_flavors(diff_files, flavors):
+    """
+    Detects changed flavors from a list of changed files and returns them as a set.
+
+    Example
+    -------
+    >>> get_changed_flavors(["mlflow/xgboost.py"], ["pytorch", "xgboost"])
+    {'xgboost'}
+    """
+    maybe_flavors = set(os.path.splitext(f.split(os.sep)[1])[0] for f in diff_files)
+    return maybe_flavors.intersection(flavors)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--diff-only",
+        action="store_true",
+        help="If specified, ignore unchanged flavors or configurations",
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    dummy_git_diff = """
+mlflow/pytorch/__init__.py
+mlflow/xgboost.py
+"""
+
+    diff_files = dummy_git_diff.strip().split("\n")
+
+    config_pr = read_yaml("ml-package-versions-pr.yml")
+    config_master = read_yaml("ml-package-versions-master.yml")
+    changed_flavors = get_changed_flavors(diff_files, set(config_pr.keys()))
+
+    names = []
+    includes = []
+    for flavor, cfgs in config_pr.items():
+        for key, cfg in cfgs.items():
+
+            if (
+                not args.diff_only
+                or (flavor in changed_flavors)
+                or (flavor not in config_master)
+                or (flavor in config_master and key not in config_master[flavor])
+                or (
+                    flavor in config_master
+                    and key in config_master[flavor]
+                    and config_master[flavor][key] != cfg
+                )
+            ):
+
+                if key == "dev":
+                    name = "-".join([flavor, key])
+                    names.append(name)
+                    includes.append(
+                        {
+                            "name": name,
+                            "flavor": flavor,
+                            "package": cfg["package"],
+                            "version": key,
+                            "install": cfg["install"],
+                            "requirements": " ".join(cfg.get("requirements", [])),
+                            "run": cfg["run"],
+                        }
+                    )
+                    continue
+
+                versions = get_latest_micro_versions(
+                    cfg["package"], cfg["minimum"], cfg["maximum"], cfg.get("unsupported")
```

Comment:
For *testing* purposes, we should test all the way up until the latest available version. We shouldn't limit ourselves to the configured maximum version in the YAML file. The goal of the YAML file is to document the latest version of each ML library that we tested MLflow against before releasing it. Accordingly, we need to test against the latest possible version in order to confidently update our `maximum` version documentation. Let me know if this makes sense.

Reply:
Makes senes. I'll fix it.